api/rest/coordinator.go

```diff
@@ -0,0 +1,231 @@
+		}
+	}
+}
+
+// NewParallelCoordinator constructs a new parallelcoordinator.
+func NewParallelCoordinator(client Client) ParallelCoordinator {
+	return ParallelCoordinator{
+		client: client,
+	}
+}
+
+// Create starts bundle creation process. Creation process could be monitor on returned channel.
+func (c ParallelCoordinator) Create(ctx context.Context, id string, nodes []node) <-chan BundleStatus {
+
+	jobs := make(chan job)
+	statuses := make(chan BundleStatus)
+
+	for i := 0; i < numberOfWorkers; i++ {
+		go worker(ctx, jobs, statuses)
+	}
+
+	for _, n := range nodes {
+		logrus.WithField("IP", n.IP).Info("Sending creation request to node.")
+
+		// necessary to prevent the closure from giving the same node to all the calls
+		tmpNode := n
+		jobs <- func() BundleStatus {
+			//TODO(janisz): Handle context
+
+			// because this will also make a request from the coordinating master, we
+			// can't tell it to make a local bundle with the same ID this guarantees that
+			// the coordinating master's local bundle has a different ID.
+			// from this point, the full <ip>-<id> will be carried in the BundleStatus ID field so this does not
+			// need to be recalculated in the coordinator
+			fullID := fmt.Sprintf("%s-%s", tmpNode.IP, id)
+			return c.createBundle(ctx, tmpNode, fullID, jobs)
+		}
+	}
+
+	return statuses
+}
+
+func (c ParallelCoordinator) Collect(ctx context.Context, bundleID string, numBundles int, statuses <-chan BundleStatus) (string, error) {
+
+	var bundles []string
+
+	for len(bundles) < numBundles {
+		select {
+		case <-ctx.Done():
+			return "", fmt.Errorf("context canceled before all node bundles finished")
+		case s := <-statuses:
+			if !s.done {
+				logrus.WithError(s.err).WithField("IP", s.node.IP).WithField("ID", s.ID).Info("Got status update. Bundle not ready.")
+				continue
+			}
+			if s.err != nil {
+				logrus.WithError(s.err).WithField("IP", s.node.IP).WithField("ID", s.ID).Warn("Bundle errored")
+				// TODO: the bundle will never finish, need to stop this from waiting forever
+				continue
+			}
+
+			//TODO(janisz): Handle file creation in workdir
+			// the '*' will be swapped out with a random string by ioutil.TempFile
+			// This will use the default temp directory from os.TempDir
+			destinationFile, err := ioutil.TempFile("", fmt.Sprintf("bundle-%s-*.zip", bundleID))
+			if err != nil {
+				return "", fmt.Errorf("could not create temporary result file")
+			}
+			_ = destinationFile.Close()
+
+			err = c.client.GetFile(ctx, s.node.baseURL, s.ID, destinationFile.Name())
+			if err != nil {
+				logrus.WithError(err).WithField("IP", s.node.IP).WithField("ID", s.ID).Warn("Could not download file")
+			}
+
+			bundles = append(bundles, destinationFile.Name())
+		}
+	}
+
+	return mergeZips(bundleID, bundles)
+}
+
+func mergeZips(bundleID string, bundlePaths []string) (string, error) {
+
+	bundleName := fmt.Sprintf("bundle-%s-*.zip", bundleID)
+	mergedZip, err := ioutil.TempFile("", bundleName)
```

Comment:
We can't do it like this. User configure workdir and we should use it. How about creating a function interface that returns `TempFile` it could be injected into coordinator and easily mocked in tests

Reply:
I would prefer to make the working directory part of the `ParallelCoordinator` struct. I don't think we need to introduce an interface and two implementations for this.

Reply:
OK, if we are going to stick with `node` we can have deterministic names for directories